src/rpp/rpp/operators/buffer.hpp

```diff
@@ -0,0 +1,113 @@
+//                  ReactivePlusPlus library
+//
+//          Copyright Aleksey Loginov 2022 - present.
+//                    TC Wang 2022 - present.
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          https://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/victimsnino/ReactivePlusPlus
+//
+
+#pragma once
+
+#include <algorithm>
+#include <mutex>
+
+#include <rpp/observables/constraints.hpp>
+#include <rpp/operators/fwd/buffer.hpp>
+#include <rpp/subscribers/constraints.hpp>
+#include <rpp/sources/create.hpp>
+
+IMPLEMENTATION_FILE(buffer_tag);
+
+namespace rpp::details
+{
+
+/// A non-copyable class that provides a copyable on_next for the subscriber and
+/// allows copies of on_next(s) to share the same states.
+template<constraint::decayed_type UpstreamType, constraint::decayed_type DownstreamType>
+struct buffer_on_next : public std::enable_shared_from_this<buffer_on_next<UpstreamType, DownstreamType>>
+{
+    using DownstreamValueType = UpstreamType;
+
+    /// \param count Number of items being bundled. Note when count == 0, we'll
+    /// treat the behavior like when count == 1.
+    explicit buffer_on_next(size_t count) {
+        buckets_.reserve(std::max<size_t>(1ul, count));
+    };
+
+    // Copy & move constructors
+    buffer_on_next(const buffer_on_next& other) = delete;
+    buffer_on_next(buffer_on_next&&) noexcept = default;
+
+    // Copy & move assignment operators
+    buffer_on_next& operator=(const buffer_on_next&) = delete;
+    buffer_on_next& operator=(buffer_on_next&&) noexcept = default;
+
+    /// \return a on_next function for subscriber, that is copyable and all the
+    /// copies share the same state.
+    auto get_on_next() {
+        return [shared_this = this->shared_from_this()](
+                const UpstreamType& value,
+                const constraint::subscriber_of_type<DownstreamType> auto& subscriber)
+        {
+            std::lock_guard<std::mutex> guard(shared_this->mutex_);
+
+            shared_this->buckets_.push_back(value);
+            if (shared_this->buckets_.size() == shared_this->buckets_.capacity())
+            {
+                auto copy = std::vector<DownstreamValueType>(shared_this->buckets_);
+
+                shared_this->buckets_.clear();
+                subscriber.on_next(std::move(copy));
```

Comment:
You can avoid extra copy
```suggestion
                subscriber.on_next(std::move(shared_this->buckets_));
                shared_this->buckets_.clear();
```

Reply:
Let's follow the discussion on the other thread.